src/type-info.c

```diff
@@ -98,36 +98,33 @@ enum vctrs_type vec_proxy_typeof(SEXP x) {
   return vec_base_typeof(x, true);
 }
 
+static bool vec_is_list_impl(SEXP proxy);
+
 // [[ include("vctrs.h") ]]
 bool vec_is_list(SEXP x) {
-  if (TYPEOF(x) != VECSXP) {
-    return false;
-  }
+  SEXP proxy = PROTECT(vec_proxy(x));
 
-  switch (class_type(x)) {
-  // Bare list
-  case vctrs_class_none:
-    return true;
+  bool out = vec_is_list_impl(proxy);
 
-  // Explicit lists
-  case vctrs_class_list:
-  case vctrs_class_list_of:
-    return true;
-
-  // Non-explicit S3 lists
-  case vctrs_class_unknown:
-    return vec_is_vector(x);
+  UNPROTECT(1);
+  return out;
+}
 
-  // TODO: Can this ever be considered a list?
-  case vctrs_class_rcrd:
+static bool vec_is_list_impl(SEXP proxy) {
```

Comment:
`vec_proxy_is_list()`? Not sure it makes sense to extract the implementation in a different funtion though.

Reply:
This was mainly to keep me from having to `UNPROTECT(1)` 3 times in `vec_is_list()` since we exit this function at different places. I don't think the name is too important here

Reply:
A standalone function conveys meaning. The name is important for this reason. I don't think we should extract logic just to avoid unprotect statements.